pkg/vsphere/tags/categories.go

```diff
@@ -0,0 +1,189 @@
+		log.Debugf("Create category failed with status code: %d, error message: %s", status, errors.ErrorStack(err))
+		return nil, errors.Errorf("Status code: %d, error: %s", status, err)
+	}
+
+	type RespValue struct {
+		Value string
+	}
+
+	var pID RespValue
+	if err := json.NewDecoder(stream).Decode(&pID); err != nil {
+		log.Debugf("Decode response body failed for: %s", errors.ErrorStack(err))
+		return nil, errors.Trace(err)
+	}
+	return &(pID.Value), nil
+}
+
+func (c *RestClient) GetCategory(ctx context.Context, id string) (*Category, error) {
+	log.Debugf("Get category %s", id)
+
+	stream, _, status, err := c.call(ctx, "GET", fmt.Sprintf("%s/id:%s", CategoryURL, id), nil, nil)
+
+	if status != http.StatusOK || err != nil {
+		log.Debugf("Get category failed with status code: %s, error message: %s", status, errors.ErrorStack(err))
+		return nil, errors.Errorf("Status code: %d, error: %s", status, err)
+	}
+
+	type RespValue struct {
+		Value Category
+	}
+
+	var pCategory RespValue
+	if err := json.NewDecoder(stream).Decode(&pCategory); err != nil {
+		log.Debugf("Decode response body failed for: %s", errors.ErrorStack(err))
+		return nil, errors.Trace(err)
+	}
+	return &(pCategory.Value), nil
+}
+
+func (c *RestClient) DeleteCategory(ctx context.Context, id string) error {
+	log.Debugf("Delete category %s", id)
+
+	_, _, status, err := c.call(ctx, "DELETE", fmt.Sprintf("%s/id:%s", CategoryURL, id), nil, nil)
+
+	if status != http.StatusOK || err != nil {
+		log.Debugf("Delete category failed with status code: %s, error message: %s", status, errors.ErrorStack(err))
+		return errors.Errorf("Status code: %d, error: %s", status, err)
+	}
+	return nil
+}
+
+func (c *RestClient) ListCategories(ctx context.Context) ([]string, error) {
+	log.Debugf("List all categories")
+
+	stream, _, status, err := c.call(ctx, "GET", CategoryURL, nil, nil)
+
+	if status != http.StatusOK || err != nil {
+		log.Debugf("Get categories failed with status code: %s, error message: %s", status, errors.ErrorStack(err))
+		return nil, errors.Errorf("Status code: %d, error: %s", status, err)
+	}
+
+	type Categories struct {
+		Value []string
+	}
+
+	var pCategories Categories
+	if err := json.NewDecoder(stream).Decode(&pCategories); err != nil {
+		log.Debugf("Decode response body failed for: %s", errors.ErrorStack(err))
+		return nil, errors.Trace(err)
+	}
+	return pCategories.Value, nil
+}
+
+func (c *RestClient) GetCategoriesByName(ctx context.Context, name string) ([]Category, error) {
+	log.Debugf("Get category %s", name)
+	categoryIds, err := c.ListCategories(ctx)
+	if err != nil {
+		log.Debugf("Get category failed for: %s", errors.ErrorStack(err))
+		return nil, errors.Trace(err)
+	}
+
+	var categories []Category
+	for _, cID := range categoryIds {
+		category, err := c.GetCategory(ctx, cID)
+		if err != nil {
+			log.Debugf("Get category %s failed for %s", cID, errors.ErrorStack(err))
+		}
+		if category.Name == name {
+			categories = append(categories, *category)
```

Comment:
Is this intended to be a deep copy? If so I'm not sure we're deep copying the arrays. If not, why not just use the reference?